nfs/src/rpc_task.cpp

```diff
@@ -318,6 +435,206 @@ void rpc_task::run_lookup()
+
+        struct write_context *cb_context = new write_context(chunk, flush_task, file_ino);
+
+        // Keep the inuse and set flag to flushing.
+        membuf->set_flushing();
+
+        do 
+        {
+            if(rpc_nfs3_write_task(flush_task->get_rpc_ctx(), write_callback, &args, cb_context) == NULL)
+            {
+                /*
+                * This call fails due to internal issues like OOM etc
+                * and not due to an actual error, hence retry.
+                */
+                rpc_retry = true;
+            }
+        } while (rpc_retry);
+    } else {
+        membuf->clear_locked();
+    }
+}
+
+void rpc_task::run_write()
+{
+    fuse_ino_t file_ino = rpc_api.write_task.get_ino();
+    auto nfs_inode = get_client()->get_nfs_inode_from_ino(file_ino);
+    size_t length = rpc_api.write_task.get_size();
+    struct fuse_bufvec* bufv = rpc_api.write_task.get_buffer_vector();
+    off_t offset = rpc_api.write_task.get_offset();
+    auto client = get_client();
+    std::vector<bytes_chunk> chunkvec;
+
+    int error_code = nfs_inode->get_write_error();
+    if (error_code < 0) {
+        AZLogDebug("Previous write to this Blob failed with error={}", error_code);
+
+        reply_error(error_code);
+        return;
+    }
+
+    chunkvec = copy_to_cache(get_client(), file_ino, bufv, offset, length, error_code);
+    assert(chunkvec.size() > 0);
+
+    // Iterate over all the chunks this write allocated.
+    for (auto &chunk : chunkvec) {
+        // Flush the membuf to backend.
+        sync_membuf(chunk, client, file_ino);
+    }
+
+    // Send reply to original request without waiting for the backend write to complete.
+    reply_write(length);
+} 
+
+void rpc_task::run_flush()
+{
+    fuse_ino_t file_ino = rpc_api.write_task.get_ino();
+    auto nfs_inode = get_client()->get_nfs_inode_from_ino(file_ino);
+
+    // Get the filecache_handle from the inode.
+    auto filecache_handle = nfs_inode->filecache_handle;
+    assert(filecache_handle != nullptr);
+
+    // Check if any write error set, If set send the response immediately.
+    int error_code = nfs_inode->get_write_error();
+    if (error_code != 0) {
+        reply_error(nfs_inode->get_write_error());
+        return;
+    }
+
+    // Get the dirty bytes_chunk from the filecache handle.
+    auto chunk_vec = filecache_handle->get_dirty_bc();
+
+    // Flush dirty membuf to backend.
+    for (auto &chunk: chunk_vec)
+    {
+        // Flush the membuf to backend.
+        sync_membuf(chunk, client, file_ino);
+    }
+
+    // Wait for membuf to flush and get result back.
+    for (auto &chunk: chunk_vec)
+    {
+        auto membuf = chunk.get_membuf();
+        assert(membuf != nullptr);
+
+        membuf->set_locked();
+        assert(membuf->is_inuse());
+
+        if (membuf->is_dirty()) {
+            assert(nfs_inode->get_write_error() != 0);
+            AZLogDebug ("Writes failed with error: {}", nfs_inode->get_write_error());
```

Comment:
technically some other thread can get the lock before we get the lock here and it may set it to dirty.